### PR TITLE
Generalize content-addressable POST request storage

### PR DESCRIPTION
**Pull request by Pchelolo**
Adding the implementation that would allow to create request chains with config. 

Notes:
1. We should provide a Template to resources now.
2. The catch condition function is created as `new Function` because it's faster than reconstructing it on each request, but I'm not sure this optimisation is needed.

PS: this PR still lacks automatic testing coverage, but I think it's better to discuss the code and then add testing for different corner cases, as these corner cases might change during the review/discussion.

Bug: https://phabricator.wikimedia.org/T105975

### CODE REVIEW

**Comment by Pchelolo** [comment]
I've benchmarked this solution yesterday with testservice, and compared to old testservice. 
- Cache-hits have ~12% performance improvement
- Cache-miss benchmarked with local http server - for 100k requests we have degradation from 500 seconds to 520 seconds, but I suppose it's acceptable.

File: mods/parsoid.js [hunk 1]
```diff
@@ -8,6 +8,7 @@ var P = require('bluebird');
 var URI = require('swagger-router').URI;
 var uuid   = require('cassandra-uuid').TimeUuid;
 var rbUtil = require('../lib/rbUtil');
+var Template = require('../lib/reqTemplate');
```
Comment (gwicke) [inline mods/parsoid.js]: Could we keep this out of the modules, and perform the Template wrapping in the module loading code in the router instead? I think that would keep the interface cleaner.

File: mods/simple_service.js [hunk 2]
```diff
@@ -19,74 +18,308 @@ function SimpleService(options) {
...
+ *
+ * @param catchDefinition the condition in the format of 'catch' and 'return_if' stanza
+ * @returns {Function} a function that verifies the condition
+ */
+function compileCatchFunction(catchDefinition) {
+    function createCondition(catchCond, option) {
+        if (catchCond === 'status') {
+            if (/^[0-9]+$/.test(option.toString())) {
+                return '(res["' + catchCond + '"] === ' + option + ')';
+            } else {
```
Comment (gwicke) [inline mods/simple_service.js]: Perhaps worth checking that the condition is actually of the form `/^[0-9]xx$/i`?

File: mods/simple_service.js [hunk 3]
```diff
@@ -19,74 +18,308 @@ function SimpleService(options) {
...
+        throw new Error('Invalid spec. ' +
+            'return_if should have a matching request. Step: ' + JSON.stringify(stepConf));
+    }
+
+    // Only supply 'catch' when 'request' is specified
+    if (Object.keys(stepConf).some(function(requestName) {
+        return stepConf[requestName].catch && !stepConf[requestName].request;
+    })) {
+        throw new Error('Invalid spec. ' +
+            'catch should have a matching request. Step: ' + JSON.stringify(stepConf));
```
Comment (gwicke) [inline mods/simple_service.js]: Hmm.. should we allow a bare catch too, so that we can set up a catch-all error handler?

Comment (Pchelolo) [inline mods/simple_service.js]: @gwicke there's a problem with bare catch: 
 - we cant's really match stuff other then a current request result, because it's implemented like this
 - if we have a catch, the request chain return undefined, which is hard to handle.

So, it's quite hard to handle, and the `catch` impl would be way slower, so I'd better not include it until we need it.

What do you think?

Comment (gwicke) [inline mods/simple_service.js]: I'm actually wondering how we'll template error responses. It seems we dropped that capability in the syntax reshuffle.

I guess the question is whether we eventually want to allow blocks like
```yaml
- return_if_error:
    return_if:
      $.earlier_request.status: 333
    return:
      body: ....
```
or, for catch,
```yaml
- catch_any_error:
    catch:
- ...
```

File: mods/simple_service.js [hunk 4]
```diff
@@ -19,74 +18,308 @@ function SimpleService(options) {
...
+        });
+    });
+}
+
+/**
+ * Prepares a request chain on startup: compiles request/response templates,
+ * creates catch conditions verifiers.
+ * @param conf request chain configuration
+ *
+ * @returns {Array} an array of prepared steps, each containing an
```
Comment (d00rman) [inline mods/simple_service.js]: *containing an* ... ?

File: mods/simple_service.js [hunk 5]
```diff
@@ -19,74 +18,308 @@ function SimpleService(options) {
...
+ */
+function prepareRequestChain(conf) {
+
+    /**
+     * Creates a function gene
```
Comment (d00rman) [inline mods/simple_service.js]: *gene* ?

File: mods/simple_service.js [hunk 6]
```diff
@@ -19,74 +18,308 @@ function SimpleService(options) {
...
+                requestSpec.request = function(restbase, context) {
+                    return template.eval(context);
+                };
+            }
+        }
+    }
+
+    function prepareCatch(requestSpec, requestConf) {
+        if (requestConf.catch) {
+            var isCatched = compileCatchFunction(requestConf.catch);
```
Comment (d00rman) [inline mods/simple_service.js]: *isCaught*

File: mods/simple_service.js (hunk 3 above)
Comment (gwicke) [inline mods/simple_service.js]: In any case, we can add this later.

**Comment by gwicke** [comment]
As a next step, I think it would be nice to move this functionality from the simple_service module into RESTBase proper. That way, we can use the same templating functionality instead of `x-backend-request` in top-level specs.

**Comment by Pchelolo** [comment]
Updated the fix according to comments. I think it's better to postpone moving whole RB on this, it should be tested better first.

**Comment by gwicke** [comment]
LGTM. @d00rman @eevans, any objections to merging this?

**Comment by d00rman** [comment]
How about adding a few tests related to this feature before merging? That could come in handy later on should any regressions arise.

**Comment by Pchelolo** [comment]
@d00rman ok, I'll add some tests. Need to cover parallel solution and validations.

**Comment by Pchelolo** [comment]
@d00rman Added tests for spec validation and parallel requests

**Comment by Pchelolo** [comment]
@d00rman Great idea with x-monitor. Forgot about this feature, fixed now.

**Comment by gwicke** [comment]
Congratulations, @Pchelolo!
